Have a more specific error message when checksum of an archive fails

Checksum errors seem to be quite a common problem with, for instance, github or others changing how archives are generated and thus changing their checksums.

Right now there seems to be no specific error message for checksum failures and we only get the same error message as when the archive couldn't be reached (i.e. `[ERROR] The sources of the following couldn't be obtained, aborting:`)

Could we have a specific error message for checksum failures ?

--- a/src/client/opamAction.ml
+++ b/src/client/opamAction.ml
@@ -154,10 +154,10 @@ let download_package st nv =
     (OpamUpdate.cleanup_source st
        (OpamPackage.Map.find_opt nv st.installed_opams)
        (OpamSwitchState.opam st nv);
-     OpamProcess.Job.catch (fun e -> Done (Some (Printexc.to_string e)))
+     OpamProcess.Job.catch (fun e -> Done (Some (None, Printexc.to_string e)))
      @@ fun () ->
      OpamUpdate.download_package_source st nv dir @@| function
-     | Some (Not_available s) -> Some s
+     | Some (Not_available (s,l)) -> Some (s,l)
      | None | Some (Up_to_date () | Result ()) -> None)
 
 (* Prepare the package build:
@@ -279,7 +279,7 @@ let prepare_package_source st nv dir =
         (OpamFile.URL.url urlf :: OpamFile.URL.mirrors urlf)
       @@| function
       | Result () | Up_to_date () -> None
-      | Not_available msg -> Some (Failure msg)
+      | Not_available (_,msg) -> Some (Failure msg)
     in
     List.fold_left (fun job dl ->
         job @@+ function

--- a/src/client/opamAction.mli
+++ b/src/client/opamAction.mli
@@ -15,12 +15,13 @@ open OpamTypes
 open OpamStateTypes
 
 (** [download t pkg] downloads the source of the package [pkg] into its locally
-    cached source dir. Returns [Some errmsg] on error, [None] on success.
+    cached source dir. Returns [Some (short_errmsg option, long_errmsg)] on error,
+    [None] on success. See {!OpamTypes.Not_available}.
 
     This doesn't update dev packages that already have a locally cached
     source. *)
 val download_package:
-  rw switch_state -> package -> string option OpamProcess.job
+  rw switch_state -> package -> (string option * string) option OpamProcess.job
 
 (** [prepare_package_source t pkg dir] updates the given source [dir] with the
     extra downloads, overlays and patches from the package's metadata

--- a/src/client/opamAdminCommand.ml
+++ b/src/client/opamAdminCommand.ml
@@ -137,7 +137,7 @@ let package_files_to_cache repo_root cache_dir ?link (nv, prefix) =
           checksums
           (OpamFile.URL.url urlf :: OpamFile.URL.mirrors urlf)
         @@| function
-        | Not_available m ->
+        | Not_available (_,m) ->
           OpamPackage.Map.update nv (fun l -> m::l) [] errors
         | Up_to_date () | Result () ->
           OpamStd.Option.iter (fun link_dir ->

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -2332,7 +2332,7 @@ let pin ?(unpin_only=false) () =
             ~cache_dir:(OpamRepositoryPath.download_cache
                           OpamStateConfig.(!r.root_dir))
             basename pin_cache_dir [] [url] @@| function
-          | Not_available u ->
+          | Not_available (_,u) ->
             OpamConsole.error_and_exit `Sync_error
               "Could not retrieve %s" u
           | Result _ | Up_to_date _ -> from_opam_files pin_cache_dir, Some cleanup
@@ -2571,7 +2571,7 @@ let source =
                (OpamPackage.to_string nv) dir []
                [url])
         with
-        | Not_available u ->
+        | Not_available (_,u) ->
           OpamConsole.error_and_exit `Sync_error "%s is not available" u
         | Result _ | Up_to_date _ ->
           OpamConsole.formatted_msg
@@ -2581,7 +2581,7 @@ let source =
       let job =
         let open OpamProcess.Job.Op in
         OpamUpdate.download_package_source t nv dir @@+ function
-        | Some (Not_available s) ->
+        | Some (Not_available (_,s)) ->
           OpamConsole.error_and_exit `Sync_error "Download failed: %s" s
         | None | Some (Result () | Up_to_date ()) ->
           OpamAction.prepare_package_source t nv dir @@| function

--- a/src/client/opamPinCommand.ml
+++ b/src/client/opamPinCommand.ml
@@ -64,7 +64,7 @@ let get_source_definition ?version st nv url =
   in
   let open OpamProcess.Job.Op in
   OpamUpdate.fetch_dev_package url srcdir nv @@| function
-  | Not_available s -> raise (Fetch_Fail s)
+  | Not_available (_,s) -> raise (Fetch_Fail s)
   | Up_to_date _ | Result _ ->
     match OpamPinned.find_opam_file_in_source nv.name srcdir with
     | None -> None

--- a/src/client/opamSolution.ml
+++ b/src/client/opamSolution.ml
@@ -368,8 +368,9 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
   if fatal_dl_error then
     OpamConsole.error_and_exit `Sync_error
       "The sources of the following couldn't be obtained, aborting:\n%s"
-      (OpamStd.Format.itemize OpamPackage.to_string
-         (OpamPackage.Map.keys failed_downloads))
+      (OpamStd.Format.itemize
+      (fun (p,s) -> Printf.sprintf "%s:\n%s" (OpamPackage.to_string p) s)
+         (OpamPackage.Map.bindings failed_downloads))
   else if not (OpamPackage.Map.is_empty failed_downloads) then
     OpamConsole.warning
       "The sources of the following couldn't be obtained, they may be \

--- a/src/client/opamSolution.ml
+++ b/src/client/opamSolution.ml
@@ -348,14 +348,14 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
     in
     List.fold_left2 (fun failed nv -> function
         | None -> failed
-        | Some e -> OpamPackage.Map.add nv e failed)
+        | Some (s,l) -> OpamPackage.Map.add nv (s,l) failed)
       OpamPackage.Map.empty sources_list results
   in
 
   if OpamClientConfig.(!r.json_out <> None) &&
      not (OpamPackage.Map.is_empty failed_downloads) then
     OpamJson.append "download-failures"
-      (`O (List.map (fun (nv,err) -> OpamPackage.to_string nv, `String err)
+      (`O (List.map (fun (nv,(_,err)) -> OpamPackage.to_string nv, `String err)
              (OpamPackage.Map.bindings failed_downloads)));
 
   let fatal_dl_error =
@@ -369,7 +369,10 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
     OpamConsole.error_and_exit `Sync_error
       "The sources of the following couldn't be obtained, aborting:\n%s"
       (OpamStd.Format.itemize
-      (fun (p,s) -> Printf.sprintf "%s:\n%s" (OpamPackage.to_string p) s)
+         (fun (p, (s,l)) ->
+            Printf.sprintf "%s:%s" (OpamPackage.to_string p)
+              (if OpamConsole.verbose () then "\n" ^ l
+               else " " ^  OpamStd.Option.default l s))
          (OpamPackage.Map.bindings failed_downloads))
   else if not (OpamPackage.Map.is_empty failed_downloads) then
     OpamConsole.warning

--- a/src/format/opamTypes.mli
+++ b/src/format/opamTypes.mli
@@ -49,7 +49,9 @@ type std_path =
 (** Download result *)
 type 'a download =
   | Up_to_date of 'a
-  | Not_available of string
+  | Not_available of string option * string
+  (** Arguments are respectively the short (if relevant) and long error message.
+      The usage is: the second argument is the default one when the first one is [None] *)
   | Result of 'a
 
 (** {2 Packages} *)

--- a/src/repository/opamDownload.ml
+++ b/src/repository/opamDownload.ml
@@ -14,6 +14,9 @@ open OpamProcess.Job.Op
 
 let log fmt = OpamConsole.log "CURL" fmt
 
+exception Download_fail of string option * string
+let fail (s,l) = raise (Download_fail (s,l))
+
 let user_agent =
   CString (Printf.sprintf "opam/%s" (OpamVersion.(to_string current)))
 
@@ -74,22 +77,25 @@ let tool_return url ret =
   match Lazy.force OpamRepositoryConfig.(!r.download_tool) with
   | _, `Default ->
     if OpamProcess.is_failure ret then
-      Printf.ksprintf failwith "Download command failed: %s"
-        (OpamProcess.result_summary ret)
+      fail (Some "Download command failed",
+                Printf.sprintf "Download command failed: %s"
+                  (OpamProcess.result_summary ret))
     else Done ()
   | _, `Curl ->
     if OpamProcess.is_failure ret then
-      Printf.ksprintf failwith "Curl failed: %s"
-        (OpamProcess.result_summary ret);
+      fail (Some "Curl failed", Printf.sprintf "Curl failed: %s"
+                  (OpamProcess.result_summary ret));
     match ret.OpamProcess.r_stdout with
     | [] ->
-      failwith (Printf.sprintf "curl: empty response while downloading %s"
+      fail (Some "curl empty response",
+                Printf.sprintf "curl: empty response while downloading %s"
                   (OpamUrl.to_string url))
     | l  ->
       let code = List.hd (List.rev l) in
       let num = try int_of_string code with Failure _ -> 999 in
       if num >= 400 then
-        failwith (Printf.sprintf "curl: code %s while downloading %s"
+        fail (Some ("curl error code " ^ code),
+                  Printf.sprintf "curl: code %s while downloading %s"
                     code (OpamUrl.to_string url))
       else Done ()
 
@@ -131,14 +137,15 @@ let really_download
   download_command ~compress ?checksum ~url ~dst:tmp_dst
   @@+ fun () ->
   if not (Sys.file_exists tmp_dst) then
-    failwith "Download command succeeded, but resulting file not found"
+    fail (None, "Download command succeeded, but resulting file not found")
   else if Sys.file_exists dst && not overwrite then
     OpamSystem.internal_error "The downloaded file will overwrite %s." dst;
   if validate &&
      OpamRepositoryConfig.(!r.force_checksums <> Some false) then
     OpamStd.Option.iter (fun cksum ->
         if not (OpamHash.check_file tmp_dst cksum) then
-          failwith (Printf.sprintf "Bad checksum, expected %s"
+          fail (Some "Bad checksum",
+                    Printf.sprintf "Bad checksum, expected %s"
                       (OpamHash.to_string cksum)))
       checksum;
   OpamSystem.mv tmp_dst dst;

--- a/src/repository/opamDownload.mli
+++ b/src/repository/opamDownload.mli
@@ -11,6 +11,8 @@
 
 (** Configuration init and handling of downloading commands *)
 
+exception Download_fail of string option * string
+
 (** downloads a file from an URL, using Curl, Wget, or a custom configured
     tool, to the given directory. Returns the downloaded filename.
     @raise Failure if the download failed or if the checksum is specified and

--- a/src/repository/opamHTTP.ml
+++ b/src/repository/opamHTTP.ml
@@ -75,13 +75,13 @@ module B = struct
     OpamProcess.Job.catch
       (fun e ->
          OpamStd.Exn.fatal e;
-         let msg =
-           Printf.sprintf "%s (%s)" (OpamUrl.to_string remote_url)
-             (match e with
-              | Failure msg -> msg
-              | _ -> "download failed")
+         let s,l =
+           let str = Printf.sprintf "%s (%s)" (OpamUrl.to_string remote_url) in
+           match e with
+           | OpamDownload.Download_fail (s,l) -> s, str l
+           | _ -> Some "Download failed", str "download failed"
          in
-         Done (Not_available msg))
+         Done (Not_available (s,l)))
     @@ fun () ->
     OpamDownload.download ~quiet:true ~overwrite:true ?checksum remote_url dirname
     @@+ fun local_file -> Done (Result (Some local_file))

--- a/src/repository/opamLocal.ml
+++ b/src/repository/opamLocal.ml
@@ -71,19 +71,19 @@ let rsync ?(args=[]) ?(exclude_vcdirs=true) src dst =
     ]
   in
   if not(remote || Sys.file_exists src) then
-    Done (Not_available src)
+    Done (Not_available (None, src))
   else if src = dst then
     Done (Up_to_date [])
   else if overlap src dst then
     (OpamConsole.error "Cannot sync %s into %s: they overlap" src dst;
-     Done (Not_available src))
+     Done (Not_available (None, src)))
   else (
     OpamSystem.mkdir dst;
     call_rsync (fun () -> not (OpamSystem.dir_is_empty dst))
       ( rsync_arg :: args @ exclude_args @
         [ "--delete"; "--delete-excluded"; src; dst; ])
     @@| function
-    | None -> Not_available src
+    | None -> Not_available (None, src)
     | Some [] -> Up_to_date []
     | Some lines -> Result lines
   )
@@ -96,13 +96,13 @@ let rsync_dirs ?args ?exclude_vcdirs url dst =
   if not (is_remote url) &&
      not (OpamFilename.exists_dir (OpamFilename.Dir.of_string src_s))
   then
-    Done (Not_available (Printf.sprintf "Directory %s does not exist" src_s))
+    Done (Not_available (None, Printf.sprintf "Directory %s does not exist" src_s))
   else
   rsync ?args ?exclude_vcdirs src_s dst_s @@| function
-  | Not_available s -> Not_available s
+  | Not_available _ as na -> na
   | Result _ ->
     if OpamFilename.exists_dir dst then Result dst
-    else Not_available dst_s
+    else Not_available (None, dst_s)
   | Up_to_date _ -> Up_to_date dst
 
 let rsync_file ?(args=[]) url dst =
@@ -110,7 +110,7 @@ let rsync_file ?(args=[]) url dst =
   let dst_s = OpamFilename.to_string dst in
   log "rsync_file src=%s dst=%s" src_s dst_s;
   if not (is_remote url || OpamFilename.(exists (of_string src_s))) then
-    Done (Not_available src_s)
+    Done (Not_available (None, src_s))
   else if src_s = dst_s then
     Done (Up_to_date dst)
   else
@@ -118,11 +118,11 @@ let rsync_file ?(args=[]) url dst =
      call_rsync (fun () -> Sys.file_exists dst_s)
        ( rsync_arg :: args @ [ src_s; dst_s ])
      @@| function
-     | None -> Not_available src_s
+     | None -> Not_available (None, src_s)
      | Some [] -> Up_to_date dst
      | Some [_] ->
        if OpamFilename.exists dst then Result dst
-       else Not_available src_s
+       else Not_available (None, src_s)
      | Some l ->
        OpamSystem.internal_error
          "unknown rsync output: {%s}"
@@ -193,7 +193,7 @@ module B = struct
     in
     rsync remote_url.OpamUrl.path dir
     @@| function
-    | Not_available d -> Not_available d
+    | Not_available _ as na -> na
     | (Result _ | Up_to_date _) as r ->
       let res x = match r with
         | Result _ -> Result x
@@ -209,7 +209,7 @@ module B = struct
       if OpamFilename.exists filename then res (Some filename)
       else
         Not_available
-          (Printf.sprintf
+          (None, Printf.sprintf
              "Could not find target file %s after rsync with %s. \
               Perhaps you meant %s/ ?"
              (OpamUrl.basename remote_url)

--- a/src/repository/opamRepository.ml
+++ b/src/repository/opamRepository.ml
@@ -77,7 +77,7 @@ let fetch_from_cache =
            else OpamConsole.colorise `red " (MISMATCH)")
           checksums);
     OpamFilename.remove file;
-    Done (Not_available "cache CONFLICT")
+    Done (Not_available (None, "cache CONFLICT"))
   in
   let dl_from_cache_job root_cache_url checksum file =
     let url = cache_url root_cache_url checksum in
@@ -89,7 +89,7 @@ let fetch_from_cache =
     | `rsync ->
       (OpamLocal.rsync_file url file @@| function
         | Result _ | Up_to_date _-> ()
-        | Not_available m -> failwith m)
+        | Not_available (_,m) -> failwith m)
     | #OpamUrl.version_control ->
       failwith "Version control not allowed as cache URL"
   in
@@ -107,13 +107,13 @@ let fetch_from_cache =
     then Done (Up_to_date (hit_file, OpamUrl.empty))
     else mismatch hit_file
   with Not_found -> match checksums with
-    | [] -> Done (Not_available "cache miss")
+    | [] -> Done (Not_available (None, "cache miss"))
     | checksum::_ ->
       (* Try all cache urls in order, but only the first checksum *)
       let local_file = cache_file cache_dir checksum in
       let tmpfile = OpamFilename.add_extension local_file "tmp" in
       let rec try_cache_dl = function
-        | [] -> Done (Not_available "cache miss")
+        | [] -> Done (Not_available (None, "cache miss"))
         | root_cache_url::other_caches ->
           OpamProcess.Job.catch
             (function Failure _ -> try_cache_dl other_caches
@@ -195,7 +195,7 @@ let pull_from_upstream
          (OpamUrl.to_string url);
        ret)
     else
-      Not_available "Checksum mismatch"
+      Not_available (None, "Checksum mismatch")
   | (Result None | Up_to_date None) as ret ->
     if checksums = [] then
       (OpamConsole.msg "[%s] %s from %s\n"
@@ -208,8 +208,8 @@ let pull_from_upstream
                           retrieved from %s"
          label (OpamUrl.to_string url);
        OpamFilename.rmdir destdir;
-       Not_available "can't check directory checksum")
-  | Not_available r -> Not_available r
+       Not_available (None, "can't check directory checksum"))
+  | Not_available _ as na -> na
 
 let rec pull_from_mirrors label ?working_dir cache_dir destdir checksums = function
   | [] -> invalid_arg "pull_from_mirrors: empty mirror list"
@@ -218,7 +218,7 @@ let rec pull_from_mirrors label ?working_dir cache_dir destdir checksums = funct
   | url::mirrors ->
     pull_from_upstream label ?working_dir cache_dir destdir checksums url
     @@+ function
-    | Not_available s ->
+    | Not_available (_,s) ->
       OpamConsole.warning "%s: download of %s failed (%s), trying mirror"
         label (OpamUrl.to_string url) s;
       pull_from_mirrors label cache_dir destdir checksums mirrors
@@ -232,9 +232,10 @@ let pull_tree
     OpamFilename.extract_job f local_dirname @@+ function
     | None -> Done (Up_to_date ())
     | Some (Failure _) ->
-      Done (Not_available (OpamFilename.(Base.to_string (basename f))^
-                           " (could not extract archive)"))
-    | Some e -> Done (Not_available (Printexc.to_string e))
+      Done (Not_available (Some "Could not extract archive",
+                           OpamFilename.(Base.to_string (basename f))^
+                            " (could not extract archive)"))
+    | Some e -> Done (Not_available (None, Printexc.to_string e))
   in
   (match cache_dir with
    | Some cache_dir ->
@@ -243,7 +244,7 @@ let pull_tree
      fetch_from_cache cache_dir cache_urls checksums
    | None ->
      assert (cache_urls = []);
-     Done (Not_available "no cache"))
+     Done (Not_available (None, "no cache")))
   @@+ function
   | Up_to_date (archive, _) ->
     OpamConsole.msg "[%s] found in cache\n"
@@ -289,7 +290,7 @@ let pull_file label ?cache_dir ?(cache_urls=[])  ?(silent_hits=false)
      fetch_from_cache cache_dir cache_urls checksums
    | None ->
      assert (cache_urls = []);
-     Done (Not_available "no cache"))
+     Done (Not_available (None, "no cache")))
   @@+ function
   | Up_to_date (f, _) ->
     if not silent_hits then
@@ -314,7 +315,7 @@ let pull_file label ?cache_dir ?(cache_urls=[])  ?(silent_hits=false)
         @@| function
         | Up_to_date _ -> assert false
         | Result (Some f) -> OpamFilename.move ~src:f ~dst:file; Result ()
-        | Result None -> Not_available "is a directory"
+        | Result None -> Not_available (None, "is a directory")
         | Not_available _ as na -> na)
 
 let pull_file_to_cache label ~cache_dir ?(cache_urls=[]) checksums remote_urls =
@@ -333,7 +334,7 @@ let pull_file_to_cache label ~cache_dir ?(cache_urls=[]) checksums remote_urls =
         @@| function
         | Up_to_date _ -> assert false
         | Result (Some _) -> Result ()
-        | Result None -> Not_available "is a directory"
+        | Result None -> Not_available (None, "is a directory")
         | Not_available _ as na -> na)
 
 let packages r =

--- a/src/repository/opamVCS.ml
+++ b/src/repository/opamVCS.ml
@@ -76,7 +76,7 @@ module Make (VCS: VCS) = struct
            (OpamFilename.Dir.to_string dirname)
            (OpamUrl.to_string url)
            (match e with Failure fw -> fw | _ -> Printexc.to_string e);
-         Done (Not_available (OpamUrl.to_string url)))
+         Done (Not_available (None, OpamUrl.to_string url)))
     @@ fun () ->
     if VCS.exists dirname then
       VCS.fetch ?cache_dir dirname url @@+ fun () ->

